New speed documentation (#7665)
PyTorchLightning/pytorch-lightning: docs/source/guides/speed.rst

.. testsetup:: *

    from pytorch_lightning.trainer.trainer import Trainer
    from pytorch_lightning.callbacks.early_stopping import EarlyStopping
    from pytorch_lightning.core.lightning import LightningModule

.. _speed:

#######################

> Speed up model training

Reviewer: What about max_time, max_epochs, max_batches?

> Speed up model training

Reviewer: We should define this does this mean training time to convergence. If so how do we define convergence? Should Torch metrics get a shoutout.

Author: Not sure what you mean here, feel free to edit

#######################

> There are multiple ways you can speed up your model.

Reviewer: This needs a little more expostion

Reviewer: Add a table of contents with short links to the relevant part of the tutorial

Reviewer: I feel like we are missing a lot of flags and tips that can help speed up convergence. The ones listed in the notion were just examples let's make sure we are being exhaustive. These should be living documents that we can update as we add new features as well so that the docs become less like long restaurant menus of features but rather more accessible tutorials.

.. _early_stopping:

**************
Early stopping
**************

.. raw:: html

    <video width="50%" max-width="400px" controls
    poster="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/trainer_flags/yt_thumbs/thumb_earlystop.png"
    src="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/yt/Trainer+flags+19-+early+stopping_1.mp4"></video>

|

Early stopping based on metric
==============================
The
:class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`
callback can be used to monitor a validation metric and stop the training when no improvement is observed.

To enable it:

- Import :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping` callback.
- Log the metric you want to monitor using :func:`~pytorch_lightning.core.lightning.LightningModule.log` method.
- Init the callback, and set `monitor` to the logged metric of your choice.
- Pass the :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping` callback to the :class:`~pytorch_lightning.trainer.trainer.Trainer` callbacks flag.

> .. code-block:: python

Reviewer: This should be as minimal as possible it should explain what early stopping is how it speeds up training convergence and link to the relevant API Reference calls. The goal of this is not to be reference itself

Author: I don't think all this info is in the API reference though, where should it be then?

   in the :class:`~pytorch_lightning.trainer.trainer.Trainer`,
   for example :paramref:`~pytorch_lightning.trainer.trainer.Trainer.check_val_every_n_epoch`
   and :paramref:`~pytorch_lightning.trainer.trainer.Trainer.val_check_interval`.
   It must be noted that the `patience` parameter counts the number of
   validation epochs with no improvement, and not the number of training epochs.
   Therefore, with parameters `check_val_every_n_epoch=10` and `patience=3`, the trainer
   will perform at least 40 training epochs before being stopped.

.. seealso::
    - :class:`~pytorch_lightning.trainer.trainer.Trainer`
    - :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`

Stopping an epoch early
=======================

You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_train_batch_start` to return ``-1`` when some condition is met.

If you do this repeatedly, for every epoch you had originally requested, then this will stop your entire run.

----------

.. _amp:

*********************************
Mixed precision (16-bit) training
*********************************

> Mixed precision is the combined use of both 32 and 16 bit floating points during model training, which reduced memory requirements and improves performance significantly, achiving over 3X speedups on modern GPUs.

Reviewer: Lets refactor this a little to something like

Mixed precision is the combined use of both 32 and 16 bit floating points to reduce memory footprint during model training, resulting in improved performance, achieving +3X speedups on modern GPUs.

Lightning offers mixed precision or 16-bit training for CPUs, GPUs, and TPUs.

.. raw:: html

    <video width="50%" max-width="400px" controls
    poster="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/trainer_flags/yt_thumbs/thumb_precision.png"
    src="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/yt/Trainer+flags+9+-+precision_1.mp4"></video>

|

16-bit precision on GPUs
========================
Mixed or 16-bit precision can cut your memory footprint by half.
If using volta architecture GPUs it can give a dramatic training speed-up as well.

When using PyTorch 1.6+, Lightning uses the native AMP implementation to support 16-bit precision.

> 

Reviewer: Let's separate the reference from exposition each explanation here should be super to the point with information to which reference documentation entry can let you do more

Author: but where should these be? the trainer?

    # DEFAULT
    trainer = Trainer(min_epochs=1, max_epochs=1000)

----------------

****************************
Control validation frequency
****************************

Check validation every n epochs
===============================
If you have a small dataset, you might want to check validation every n epochs. Use the `check_val_every_n_epoch` Trainer flag.

.. testcode::

    # DEFAULT
    trainer = Trainer(check_val_every_n_epoch=1)

Set validation check frequency within 1 training epoch
======================================================
For large datasets, it's often desirable to check validation multiple times within a training loop.
Pass in a float to check that often within 1 training epoch. Pass in an int `k` to check every `k` training batches.
Must use an `int` if using an `IterableDataset`.

.. testcode::

    # DEFAULT
    trainer = Trainer(val_check_interval=0.95)

    # check every .25 of an epoch
    trainer = Trainer(val_check_interval=0.25)

    # check every 100 train batches (ie: for `IterableDatasets` or fixed frequency)
    trainer = Trainer(val_check_interval=100)

----------------

******************
Limit dataset size
******************

> Use data subset for training, validation, and test

Reviewer: Explain how this leads to faster convergence

Author: I'm assuming large datasets but idk, feel free to add

.. testsetup:: *

    from pytorch_lightning.trainer.trainer import Trainer
    from pytorch_lightning.callbacks.early_stopping import EarlyStopping
    from pytorch_lightning.core.lightning import LightningModule

.. _speed:

#######################
Speed up model training
#######################

There are multiple ways you can speed up your model's time to convergence:

* `<Early stopping_>`_

* `<GPU/TPU training_>`_

* `<Mixed precision (16-bit) training_>`_

* `<Control Training Epochs_>`_
* `<Control Validation Frequency_>`_

* `<Limit Dataset Size_>`_

.. _early_stopping:

**************
Early stopping

> **************

Reviewer 2: there seems to be a misunderstanding. 
early stopping is a regularization technique. it does not decrease the time the model takes to converge. 

Here is a list of real ways one can increase convergence speed :) 
- changes to the optimization algorithm (different optimizer, different learning rate)
- different model hyperparameters (hidden size, number of layers)
- special normalization layers (batch norm, instance norm, etc.)
- different initialization

Author: Removed early stopping for now till we find the right place for it. 
Please feel free to add to the PR your suggestion for more ways to speed!

**Use when:**

.. raw:: html

    <video width="50%" max-width="400px" controls
    poster="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/trainer_flags/yt_thumbs/thumb_earlystop.png"
    src="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/yt/Trainer+flags+19-+early+stopping_1.mp4"></video>

|

> Early stopping is an optimization technique used to avoid overfitting without compromising the model's accuracy. When training a large network, there will be a point during training when the model will stop generalizing and start learning the statistical noise in the training dataset. To avoid overfitting and reduce training time, unable early stopping to stop training at the point when performance on a validation dataset starts to degrade (you can pick the metric to monitor).

Reviewer 2: ```suggestion
Early stopping is an optimization technique used to avoid overfitting without compromising the model's accuracy. When training a large network, there will be a point during training when the model will stop generalizing and start learning the statistical noise in the training dataset. To avoid overfitting and reduce training time, enable early stopping to stop training at the point when performance on a validation dataset starts to degrade (you can pick the metric to monitor).
```

Stop training when metric is degrading
======================================
The
:class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`

> callback can be used to monitor a validation metric and stop the training when no improvement is observed.

Reviewer 2: ```suggestion
callback can be used to monitor a (validation) metric and stop the training when no improvement is observed.
```
it does not have to be related to the validation loop, most of the time it is, but it could simply also be the training loss or whatever.

You can also set `check_finite` to stop training when the monitored metric becomes NaN or infinite.

Learn more in the :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping` doc.

Stopping an epoch early
=======================

You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.hooks.ModelHooks.on_train_batch_start` to return ``-1`` when some condition is met.

If you do this repeatedly, for every epoch you had originally requested, then this will stop your entire run.

----------

****************
GPU/TPU training
****************

**Use when:** Running large datasets or want to speed up your training.

With Lightning, running on GPUs, TPUs or multiple node is a simple switch of a flag.

GPU training
============

Lightning supports a variety of plugins to further speed up distributed GPU training. Most notably:

* :class:`~pytorch_lightning.plugins.training_type.DDPPlugin`
* :class:`~pytorch_lightning.plugins.training_type.DDPShardedPlugin`
* :class:`~pytorch_lightning.plugins.training_type.DeepSpeedPlugin`

.. code-block:: python

    # run on 1 gpu

>     trainer = Trainer(gpus=1)

Reviewer 2: `accelerator="ddp"` is not mentioned here, but considering this document is about speed then ddp is definitely the preferred default plugin and I think should be mentioned.

Author: is it not the default?

2. ``num_workers=1`` means ONLY one worker (just not the main process) will load data but it will still be slow.
3. The ``num_workers`` depends on the batch size and your machine.
4. A general place to start is to set ``num_workers`` equal to the number of CPUs on that machine.

.. warning:: Increasing ``num_workers`` will ALSO increase your CPU memory consumption.

The best thing to do is to increase the ``num_workers`` slowly and stop once you see no more improvement in your training speed.

Spawn
"""""
When using ``accelerator=ddp_spawn`` or training on TPUs, the way multiple GPUs/TPU cores are used is by calling ``.spawn()`` under the hood.
The problem is that PyTorch has issues with ``num_workers > 0`` when using ``.spawn()``. For this reason we recommend you
use ``accelerator=ddp`` so you can increase the ``num_workers``, however your script has to be callable like so:

.. code-block:: bash

    python my_program.py --gpus X

TPU training
============

.. code-block:: python

    # train on 1 TPU core
    trainer = Trainer(tpu_cores=1)

    # train on 8 TPU cores
    trainer = Trainer(tpu_cores=8)

To train on more than 8 cores (ie: a POD),

> submit this script using the xla_dist script.

Reviewer 3: ```suggestion
run this script using the xla_dist launcher.
```

.. _amp:

*********************************
Mixed precision (16-bit) training
*********************************

**Use when:**

* You want to optimize for memory usage on a GPU.
* You have a GPU that supports 16 bit precision.
* Your optimization algorithm (training_step) is numerically stable.
* You want to be the cool guy in the lab :p

.. raw:: html

    <video width="50%" max-width="400px" controls
    poster="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/trainer_flags/yt_thumbs/thumb_precision.png"
    src="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/yt/Trainer+flags+9+-+precision_1.mp4"></video>

|

Mixed precision combines the use of both 32 and 16 bit floating points to reduce memory footprint during model training, resulting in improved performance, achieving +3X speedups on modern GPUs.

Lightning offers mixed precision or 16-bit training for CPUs, GPUs, and TPUs. 

.. testcode::
    :skipif: not _APEX_AVAILABLE and not _NATIVE_AMP_AVAILABLE or not torch.cuda.is_available()

    # 16-bit precision
    trainer = Trainer(precision=16, gpus=4)

----------------

***********************
Control Training Epochs
***********************

> **Use when:** when you run a hyperparameter search to find good initial parameters and want to save cost (money) or power (environment).
> It can allow you to be more cost efficient and also run more experiments at the same time.

Reviewer 3: ```suggestion
**Use when:** when you wish to control the training process by epochs, steps, duration, and many other flags.
It could assist you to be efficient and run more experiments at the same time.
```

******************
Limit Dataset Size
******************

Use data subset for training, validation, and test
==================================================

**Use when:** Debugging or running huge datasets.

If you don't want to check 100% of the training/validation/test set set these flags:

.. testcode::

    # DEFAULT
    trainer = Trainer(
        limit_train_batches=1.0,
        limit_val_batches=1.0,
        limit_test_batches=1.0
    )

    # check 10%, 20%, 30% only, respectively for training, validation and test set
    trainer = Trainer(
        limit_train_batches=0.1,
        limit_val_batches=0.2,
        limit_test_batches=0.3
    )

If you also pass ``shuffle=True`` to the dataloader, a different random subset of your dataset will be used for each epoch; otherwise the same subset will be used for all epochs.

.. note:: ``limit_train_batches``, ``limit_val_batches`` and ``limit_test_batches`` will be overwritten by ``overfit_batches`` if ``overfit_batches`` > 0. ``limit_val_batches`` will be ignored if ``fast_dev_run=True``.

.. note:: If you set ``limit_val_batches=0``, validation will be disabled.

Learn more in our :ref:`trainer_flags` guide.

-----

> *********************
> Preload Data Into RAM
> *********************
> 
> When your training or preprocessing requires many operations to be performed on entire dataset(s), it can

Reviewer 2: ```suggestion
*********************
Preload Data Into RAM
*********************

**Use when:** you need access to all samples in a dataset at once. 

When your training or preprocessing requires many operations to be performed on entire dataset(s), it can
```

adding a "use when" idea here :) 
(I wrote these docs)

            g_X = self.sample_G(batch_size)

            ##########################
            # Optimize Discriminator #
            ##########################
            with d_opt.toggle_model(sync_grad=is_last_batch_to_accumulate):
                d_x = self.D(X)
                errD_real = self.criterion(d_x, real_label)

                d_z = self.D(g_X.detach())
                errD_fake = self.criterion(d_z, fake_label)

                errD = (errD_real + errD_fake)

                self.manual_backward(errD)
                if is_last_batch_to_accumulate:
                    d_opt.step()
                    d_opt.zero_grad()

            ######################
            # Optimize Generator #
            ######################
            with g_opt.toggle_model(sync_grad=is_last_batch_to_accumulate):
                d_z = self.D(g_X)
                errG = self.criterion(d_z, real_label)

                self.manual_backward(errG)
                if is_last_batch_to_accumulate:
                    g_opt.step()
                    g_opt.zero_grad()

            self.log_dict({'g_loss': errG, 'd_loss': errD}, prog_bar=True)

-----

*****************

> Set Grads to None

Reviewer 3: TMI for this guide, it shouldn't be in `speed`

*****************

In order to modestly improve performance, you can override :meth:`~pytorch_lightning.core.lightning.LightningModule.optimizer_zero_grad`.

For a more detailed explanation of pros / cons of this technique,
read `this <https://pytorch.org/docs/master/optim.html#torch.optim.Optimizer.zero_grad>`_ documentation by the PyTorch team.

.. testcode::

    class Model(LightningModule):

        def optimizer_zero_grad(self, epoch, batch_idx, optimizer, optimizer_idx):
            optimizer.zero_grad(set_to_none=True)

-----

***************

> Things to avoid

Reviewer 3: This section shouldn't be in the `speed` guide?

Author: This are things that will increase training time, but might be otherwise useful in some cases? idk

Reviewer 3: Nvm! Sounds good.

.. testsetup:: *

    from pytorch_lightning.trainer.trainer import Trainer
    from pytorch_lightning.callbacks.early_stopping import EarlyStopping
    from pytorch_lightning.core.lightning import LightningModule

.. _speed:

#######################
Speed up model training
#######################

There are multiple ways you can speed up your model's time to convergence:

* `<GPU/TPU training_>`_

* `<Mixed precision (16-bit) training_>`_

* `<Control Training Epochs_>`_

* `<Control Validation Frequency_>`_

* `<Limit Dataset Size_>`_

* `<Preload Data Into RAM_>`_

* `<Model Toggling_>`_

* `<Set Grads to None_>`_

* `<Things to avoid_>`_

****************
GPU/TPU training
****************

> **Use when:** Running large datasets or want to speed up your training.

Reviewer 4: Should we change this to something like "whenever possible"? Personally, I only train on CPU if it doesn't fit into GPU memory (extremly large data)

GPU training
============

Lightning supports a variety of plugins to further speed up distributed GPU training. Most notably:

* :class:`~pytorch_lightning.plugins.training_type.DDPPlugin`
* :class:`~pytorch_lightning.plugins.training_type.DDPShardedPlugin`
* :class:`~pytorch_lightning.plugins.training_type.DeepSpeedPlugin`

.. code-block:: python

    # run on 1 gpu
    trainer = Trainer(gpus=1)

    # train on 8 gpus, using DDP plugin
    trainer = Trainer(gpus=8, accelerator="ddp")

    # train on multiple GPUs across nodes (uses 8 gpus in total)
    trainer = Trainer(gpus=2, num_nodes=4)

GPU Training Speedup Tips
-------------------------

When training on single or multiple GPU machines, Lightning offers a host of advanced optimizations to improve throughput, memory efficiency, and model scaling.
Refer to :doc:`Advanced GPU Optimized Training for more details <../advanced/advanced_gpu>`.

Prefer DDP over DP
^^^^^^^^^^^^^^^^^^
:class:`~pytorch_lightning.plugins.training_type.DataParallelPlugin` performs three GPU transfers for EVERY batch:

1. Copy model to device.
2. Copy data to device.
3. Copy outputs of each device back to master.

> |

Reviewer 5: ```suggestion

```

> Whereas :class:`~pytorch_lightning.plugins.training_type.DDPPlugin` only performs 1 transfer to sync gradients, making DDP MUCH faster than DP.

Reviewer 5: Do we know in general how much fast DDP is than DP (factor 2, 5, 10)?

Author: i don't think we do. @awaelchli or @SeanNaren maybe?

Author: from @awaelchli it's less than X2, so I think we can skip for now

When using DDP set find_unused_parameters=False
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
By default we have set ``find_unused_parameters`` to True for compatibility issues that have arisen in the past (see the `discussion <https://github.com/PyTorchLightning/pytorch-lightning/discussions/6219>`_ for more information).

> This by default comes with a performance hit, and can be disabled in most cases.

Reviewer 6: ```suggestion
This comes with a performance hit, and can be disabled in most cases.
```

.. code-block:: python

    from pytorch_lightning.plugins import DDPPlugin

    trainer = pl.Trainer(
        gpus=2,
        plugins=DDPPlugin(find_unused_parameters=False),
    )

Dataloaders
^^^^^^^^^^^
When building your DataLoader set ``num_workers > 0`` and ``pin_memory=True`` (only for GPUs).

.. code-block:: python

    Dataloader(dataset, num_workers=8, pin_memory=True)

num_workers
"""""""""""

The question of how many workers to specify in ``num_workers`` is tricky. Here's a summary of
some references, [`1 <https://discuss.pytorch.org/t/guidelines-for-assigning-num-workers-to-dataloader/813>`_], and our suggestions:

1. ``num_workers=0`` means ONLY the main process will load batches (that can be a bottleneck).
2. ``num_workers=1`` means ONLY one worker (just not the main process) will load data but it will still be slow.
3. The ``num_workers`` depends on the batch size and your machine.

> 4. A general place to start is to set ``num_workers`` equal to the number of CPUs on that machine.

Reviewer 5: ```suggestion
4. A general place to start is to set ``num_workers`` equal to the number of CPU cores on that machine. You can get the number of CPU cores in python using `os.cpu_count()`.
```

> 4. A general place to start is to set ``num_workers`` equal to the number of CPU cores on that machine. You can get the number of CPU cores in python using `os.cpu_count()`.

Reviewer 7: ```suggestion
4. A general place to start is to set ``num_workers`` equal to the number of CPU cores on that machine. You can get the number of CPU cores in python using `os.cpu_count()`, but note that depending on your batch size, you may overflow RAM memory.
```

> .. warning:: Increasing ``num_workers`` will ALSO increase your CPU memory consumption.

Reviewer 4: Don't have a suggestion for the phrasing, but maybe add on how much it will be increased ( ~1x memory footprint of dataset / worker)?

Author: Do we have the number?

Reviewer 4: Not exactly. the main issue is how large your dataset will be in memory since that one is duplicated per worker

The best thing to do is to increase the ``num_workers`` slowly and stop once you see no more improvement in your training speed.

Spawn
"""""
When using ``accelerator=ddp_spawn`` or training on TPUs, the way multiple GPUs/TPU cores are used is by calling ``.spawn()`` under the hood.

> The problem is that PyTorch has issues with ``num_workers > 0`` when using ``.spawn()``. For this reason we recommend you

Reviewer 4: maybe also add this to num_workers section? I wouldn't read here if I have to use spawn and want to tune the workers.

Author: You mean the section above? Should we change the title?

Reviewer 4: Yes, I mean the section above. How would you change the title to reflect them both?

use ``accelerator=ddp`` so you can increase the ``num_workers``, however your script has to be callable like so:

.. code-block:: bash

>     python my_program.py --gpus X

Reviewer 4: ```suggestion
    python my_program.py
```

this implies, that the script needs a `--gpus` flag, isn't necessary for ddp to work though

TPU training
============

.. code-block:: python

    # train on 1 TPU core
    trainer = Trainer(tpu_cores=1)

    # train on 8 TPU cores
    trainer = Trainer(tpu_cores=8)

> 

Reviewer 5: Maybe note here that the parameter cannot be anything other than 1 or 8?

To train on more than 8 cores (ie: a POD),
submit this script using the xla_dist script.

Example::

    python -m torch_xla.distributed.xla_dist
    --tpu=$TPU_POD_NAME
    --conda-env=torch-xla-nightly
    --env=XLA_USE_BF16=1
    -- python your_trainer_file.py

Read more in our :ref:`accelerators` and :ref:`plugins` guides.

-----------

.. _amp:

*********************************
Mixed precision (16-bit) training
*********************************

**Use when:**

* You want to optimize for memory usage on a GPU.

> * You have a GPU that supports 16 bit precision.

Reviewer 5: ```suggestion
* You have a GPU that supports 16 bit precision (nvidia pascal architecture or newer).
```

* Your optimization algorithm (training_step) is numerically stable.

> * You want to be the cool guy in the lab :p

Reviewer 6: Maybe gender neutral is better here:
```suggestion
* You want to be the cool person in the lab :p
```

Author: Definetly.

.. raw:: html

    <video width="50%" max-width="400px" controls
    poster="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/trainer_flags/yt_thumbs/thumb_precision.png"
    src="https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/yt/Trainer+flags+9+-+precision_1.mp4"></video>

|

Mixed precision combines the use of both 32 and 16 bit floating points to reduce memory footprint during model training, resulting in improved performance, achieving +3X speedups on modern GPUs.

> Lightning offers mixed precision or 16-bit training for CPUs, GPUs, and TPUs.

Reviewer 5: We directly have a note in `trainer.rst` file saying
"""
.. note:: 16-bit precision is not supported on CPUs.
"""
so one of them is clearly wrong :P

.. testcode::
    :skipif: not _APEX_AVAILABLE and not _NATIVE_AMP_AVAILABLE or not torch.cuda.is_available()

    # 16-bit precision
    trainer = Trainer(precision=16, gpus=4)

----------------

***********************
Control Training Epochs
***********************

> **Use when:** when you run a hyperparameter search to find good initial parameters and want to save cost (money) or power (environment).

Reviewer 6: ```suggestion
**Use when:** You run a hyperparameter search to find good initial parameters and want to save cost (money) or power (environment).
```

> **Use when:** when you run a hyperparameter search to find good initial parameters and want to save cost (money) or power (environment).

Reviewer 5: ```suggestion
**Use when:** when you run a hyperparameter search to find good initial parameters and want to save time, cost (money) or power (environment).
```

It can allow you to be more cost efficient and also run more experiments at the same time.

You can use Trainer flags to force training for a minimum number of epochs or limit to a max number of epochs. Use the `min_epochs` and `max_epochs` Trainer flags to set the number of epochs to run.

.. testcode::

    # DEFAULT
    trainer = Trainer(min_epochs=1, max_epochs=1000)

> If running iteration based training, i.e. infitine / iterable dataloader, you can also control the number of steps with the `min_steps` and  `max_steps` flags:

Reviewer 5: ```suggestion
If running iteration based training, i.e. infinite / iterable dataloader, you can also control the number of steps with the `min_steps` and  `max_steps` flags:
```

.. testcode::

    trainer = Trainer(max_steps=1000)

    trainer = Trainer(min_steps=100)

You can also interupt training based on training time:

.. testcode::

    # Stop after 12 hours of training or when reaching 10 epochs (string)
    trainer = Trainer(max_time="00:12:00:00", max_epochs=10)

    # Stop after 1 day and 5 hours (dict)
    trainer = Trainer(max_time={"days": 1, "hours": 5})

Learn more in our :ref:`trainer_flags` guide.

----------------

****************************
Control Validation Frequency
****************************

Check validation every n epochs
===============================

> **Use when:** You have a small dataset, and want to run less validation checks.

Reviewer 5: ```suggestion
**Use when:** You have a small training dataset, and want to run less validation checks.
```

You can limit validation check to only run every n epochs using the `check_val_every_n_epoch` Trainer flag.

.. testcode::

    # DEFAULT
    trainer = Trainer(check_val_every_n_epoch=1)

Set validation check frequency within 1 training epoch
======================================================

> **Use when:** You have a large dataset, and want to run mid-epoch validation checks.

Reviewer 5: ```suggestion
**Use when:** You have a large training dataset, and want to run mid-epoch validation checks.
```

For large datasets, it's often desirable to check validation multiple times within a training loop.
Pass in a float to check that often within 1 training epoch. Pass in an int `k` to check every `k` training batches.
Must use an `int` if using an `IterableDataset`.

.. testcode::

    # DEFAULT
    trainer = Trainer(val_check_interval=0.95)

    # check every .25 of an epoch
    trainer = Trainer(val_check_interval=0.25)

    # check every 100 train batches (ie: for `IterableDatasets` or fixed frequency)
    trainer = Trainer(val_check_interval=100)

Learn more in our :ref:`trainer_flags` guide.

----------------

******************
Limit Dataset Size
******************

> Use data subset for training, validation, and test

Reviewer 6: ```suggestion
Use data subset for training, validation, and test.
```

==================================================

**Use when:** Debugging or running huge datasets.

If you don't want to check 100% of the training/validation/test set set these flags:

.. testcode::

    # DEFAULT
    trainer = Trainer(
        limit_train_batches=1.0,
        limit_val_batches=1.0,
        limit_test_batches=1.0
    )

    # check 10%, 20%, 30% only, respectively for training, validation and test set
    trainer = Trainer(
        limit_train_batches=0.1,
        limit_val_batches=0.2,
        limit_test_batches=0.3
    )

If you also pass ``shuffle=True`` to the dataloader, a different random subset of your dataset will be used for each epoch; otherwise the same subset will be used for all epochs.

.. note:: ``limit_train_batches``, ``limit_val_batches`` and ``limit_test_batches`` will be overwritten by ``overfit_batches`` if ``overfit_batches`` > 0. ``limit_val_batches`` will be ignored if ``fast_dev_run=True``.

.. note:: If you set ``limit_val_batches=0``, validation will be disabled.

Learn more in our :ref:`trainer_flags` guide.

-----

*********************
Preload Data Into RAM
*********************

> **Use when:** you need access to all samples in a dataset at once.

Reviewer 6: ```suggestion
**Use when:** You need access to all samples in a dataset at once.
```

When your training or preprocessing requires many operations to be performed on entire dataset(s), it can
sometimes be beneficial to store all data in RAM given there is enough space.
However, loading all data at the beginning of the training script has the disadvantage that it can take a long
time and hence it slows down the development process. Another downside is that in multiprocessing (e.g. DDP)
the data would get copied in each process.
One can overcome these problems by copying the data into RAM in advance.
Most UNIX-based operating systems provide direct access to tmpfs through a mount point typically named ``/dev/shm``.

0.  Increase shared memory if necessary. Refer to the documentation of your OS how to do this.

1.  Copy training data to shared memory:

    .. code-block:: bash

        cp -r /path/to/data/on/disk /dev/shm/

2.  Refer to the new data root in your script or command line arguments:

    .. code-block:: python

        datamodule = MyDataModule(data_root="/dev/shm/my_data")

---------

**************
Model Toggling
**************

> **Use when:** performing gradient accumulation with multiple optimizers in a

Reviewer 6: ```suggestion
**Use when:** Performing gradient accumulation with multiple optimizers in a
```

distributed setting.

Here is an explanation of what it does:

* Considering the current optimizer as A and all other optimizers as B.
* Toggling means that all parameters from B exclusive to A will have their ``requires_grad`` attribute set to ``False``.
* Their original state will be restored when exiting the context manager.

When performing gradient accumulation, there is no need to perform grad synchronization during the accumulation phase.
Setting ``sync_grad`` to ``False`` will block this synchronization and improve your training speed.

:class:`~pytorch_lightning.core.optimizer.LightningOptimizer` provides a
:meth:`~pytorch_lightning.core.optimizer.LightningOptimizer.toggle_model` function as a
:func:`contextlib.contextmanager` for advanced users.

Here is an example for advanced use-case:

> .. testcode:: python

Reviewer 7: ```suggestion
.. testcode::
```

.. testsetup:: *

>     from pytorch_lightning.trainer.trainer import Trainer
>     from pytorch_lightning.callbacks.early_stopping import EarlyStopping
>     from pytorch_lightning.core.lightning import LightningModule

Reviewer 3: ```suggestion
    from pytorch_lightning import LightningModule, Trainer
    from pytorch_lightning.callbacks import EarlyStopping
```
